sources/appendix-value-extraction.asciidoc

```diff
@@ -0,0 +1,437 @@
+----
+
+[source,java]
+----
+public class Example {
+    private CrazyMap<@RegExp(...) String, @Min(0) Long> crazyMap = ...;
+}
+----
+
+Assuming there is no dedicated extractor for `CrazyMap` but only the default ones for `K` and `V` of `Map`, extraction for `@RegExp` must happen via the default map value extractor and extraction for `@Min` via the default map key extractor.
+
+A type parameter in a sub-type may also map to several type parameters in a super-type:
+
+[source,java]
+----
+interface NumericMap<T extends Number> extends Map<T, T> {}
+----
+
+[source,java]
+----
+private NumericMap<@Min(1) Integer> integerMap;
+----
+
+The `@Min` constraint is to be applied to the map's keys and values as the annotated type parameter maps to `K` and `V` of `Map`.
+
+* Type-level constraints can be applied to the elements of `Object` arrays and arrays of any primitive type:
+
+[source,java]
+----
+String @Email[] emails;
+----
+
+[source,java]
+----
+int @Min(1) [] positiveNumbers;
+----
+
+* The extractor for cascaded validation is determined based on an element's runtime type:
+
+[source,java]
+----
+Collection<@Valid Order> orders = new ArrayList<>();
+----
+
+Here the most-specific extractor for the runtime type `ArrayList` must be applied, causing the property nodes of violations to have an index set (`Node#getIndex()`).
+
+* The container value passed to a value extractor is retrieved from the element annotated with the constraint or `@Valid`:
+
+[source,java]
+----
+private Map<String, @Valid @RetailOrder Order> ordersByName;
+
+public get Map<@NotNull String, Order> getOrdersByName() {
+    return ordersByName;
+}
+----
+
+When validating the `@NotNull` constraint, the map as returned by the getter will be passed to the map key extractor in order to obtain the map keys. When validating the `@RetailOrder` constraint and performing cascaded validation, the map as obtained directly from the field will be passed to the map value extractor in order to obtain the map values.
+
+* Custom extractor for a `Tuple` type:
+
+[source,java]
+----
+public interface Tuple<T1, T2> {
+    T1 getFirst();
+    T2 getSecond();
+}
+----
+
+[source,java]
+----
+public class TupleFirstExtractor implements ValueExtractor<Tuple<@ExtractedValue ?, ?>> {
+
+	@Override
+	public void extractValues(Tuple<?, ?> originalValue, ValueReceiver receiver) {
+		receiver.value( originalValue.getFirst(), "<first>" );
+		}
+	}
+}
+----
+
+[source,java]
+----
+public class TupleSecondExtractor implements ValueExtractor<Tuple<?, @ExtractedValue ?>> {
+
+	@Override
+	public void extractValues(Tuple<?, ?> originalValue, ValueReceiver receiver) {
+		receiver.value( originalValue.getSecond(), "<second>" );
+		}
+	}
+}
+----
+
+[source,java]
+----
+private Tuple<@NotNull @Email String, @NotNull @Min(1) Integer> tuple;
+----
+
+==== Invalid examples
+
+* Extractor types must declare exactly one parameter as the extracted value. Hence this extractor type is invalid:
```

Comment:
is this rule explicit in the formal part of the proposal. I don't think I read it.